subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java

```diff
@@ -32,82 +34,99 @@
+        resourceFilter.appendConfigurationToHasher(hasher);
     }
 
     @Nullable
-    private HashCode hashJarContents(RegularFileSnapshot jarFileSnapshot) {
+    private HashCode hashZipContents(RegularFileSnapshot zipFileSnapshot) {
         try {
-            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(jarFileSnapshot.getAbsolutePath());
+            List<FileSystemLocationFingerprint> fingerprints = fingerprintZipEntries(zipFileSnapshot.getAbsolutePath());
             if (fingerprints.isEmpty()) {
                 return null;
             }
             Hasher hasher = Hashing.newHasher();
             NormalizedPathFingerprintCompareStrategy.appendSortedToHasher(hasher, fingerprints);
             return hasher.hash();
         } catch (Exception e) {
-            return hashMalformedZip(jarFileSnapshot, e);
+            return hashMalformedZip(zipFileSnapshot, e);
         }
     }
 
-    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String jarFile) throws IOException {
-        List<FileSystemLocationFingerprint> fingerprints = Lists.newArrayList();
-        InputStream fileInputStream = null;
+    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String zipFile) throws IOException {
+        ZipInput input = null;
         try {
-            fileInputStream = Files.newInputStream(Paths.get(jarFile));
-            ZipInputStream zipInput = new ZipInputStream(fileInputStream);
-            ZipEntry zipEntry;
-            RelativePathFactory relativePathFactory = new RelativePathFactory();
-
-            while ((zipEntry = zipInput.getNextEntry()) != null) {
-                relativePathFactory.setZipEntry(zipEntry);
-                if (zipEntry.isDirectory() || classpathResourceFilter.shouldBeIgnored(relativePathFactory)) {
-                    continue;
-                }
-                HashCode hash = classpathResourceHasher.hash(zipEntry, zipInput);
+            input = new FileZipInput(new File(zipFile));
+            return fingerprintZipEntries("", input);
+        } finally {
+            IoActions.closeQuietly(input);
+        }
+    }
+
+    private List<FileSystemLocationFingerprint> fingerprintZipEntries(String parentName, ZipInput input) throws IOException {
+        List<FileSystemLocationFingerprint> fingerprints = Lists.newArrayList();
+        RelativePathFactory relativePathFactory = new RelativePathFactory();
+
+        for (ZipEntry zipEntry : input) {
+            relativePathFactory.setZipEntry(zipEntry);
+            if (zipEntry.isDirectory() || resourceFilter.shouldBeIgnored(relativePathFactory)) {
+                continue;
+            }
+            String fullName = parentName.isEmpty() ? zipEntry.getName() : parentName + "/" + zipEntry.getName();
+            if (isZipFile(zipEntry.getName())) {
+                fingerprints.addAll(fingerprintZipEntries(fullName, new StreamZipInput(zipEntry.getInputStream())));
```

Comment:
Passing the same `fingerprints` list around would allow for fewer allocations.

Reply:
Yes, this is what @paplorinc already noted above. And passing a TreeSet around would avoid sorting afterwards. I can make the change now if you want, though it will have very little impact on real world performance I suspect.

Reply:
We can leave this like it is.